metals/src/main/scala/scala/meta/internal/metals/CompilerConfiguration.scala

```diff
@@ -116,24 +120,29 @@ class CompilerConfiguration(
             presentationCompilerFuture,
             Duration(compilerConfig.timeoutDelay, compilerConfig.timeoutUnit),
           )
-          presentationCompilerRef.set(result)
           result
         }
       } catch {
         case _: ju.concurrent.TimeoutException =>
           scribe.warn(
             s"Still waiting for information about classpath, using standalone compiler for now"
           )
-          val pc = presentationCompilerRef.updateAndGet { old =>
+          this.synchronized {
+            val old = presentationCompilerRef.get()
             if (old != null) old
-            else if (!wasResolved.get()) fallback
-            else null
+            else {
+              val newFallback = fallback
+              presentationCompilerRef.set(newFallback)
+              newFallback
+            }
           }
-          if (pc != null) pc
-          // null is only returned if classpath was resolved while we were waiting and there was a race
-          else await
       }
     }
+
+    def shutdown(): Unit = {
+      cancelCompilerPromise.trySuccess(())
+      presentationCompilerRef.get().shutdown()
```

Comment:
If cancel doesn't work this doesn't really fix the issue. How about adding:
```Scala
presentationCompilerFuture.onComplete {
    case Success(pc) => pc.shutdown()
    case _ =>
  }
```
I think `pc.shutdown()` is idempotent.

Reply:
> If cancel doesn't work 

why would cancel not work?

Reply:
If getting classpath completed but before `presentationCompilerRef` is set. Also can't `presentationCompilerRef.get()` be `null`?

Reply:
makes sense, changed it according to your suggestions! And also wrapped in an option